### PR TITLE
Fix bug with gathering ports for NEG.

### PR DESCRIPTION
**Pull request by alan-kut**

### CODE REVIEW

File: pkg/controller/translator/translator.go [hunk 1]
```diff
@@ -464,10 +465,15 @@ func (t *Translator) GatherEndpointPorts(svcPorts []utils.ServicePort) []string
 			// TODO(mixia): refactor firewall syncing into a separate go routine with different trigger.
 			// With NEG, endpoint changes may cause firewall ports to be different if user specifies inconsistent backends.
 			var endpointPorts []int
+			targetPort, err := getTargetPortFromService(t.ctx.ServiceInformer.GetIndexer(), p.ID.Service.Namespace, p.ID.Service.Name, p.TargetPort)
```
[the remark this answers is not in the document]
Comment (alan-kut) [reply pkg/controller/translator/translator.go:468]: You are right. Updated.

So I guess the correct algorithm is:
- if ingress port is named -> find a port in a service by a name
- if ingress port is a number -> find a port in a service by a number
then in the found port
- if a target port is a number -> use this number
- if a target port is a name -> find the port in endpoints by the name of the port, not a target name of the port

File: pkg/controller/translator/translator.go [hunk 2]
```diff
@@ -556,6 +562,52 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
...
+	if !exists {
+		return "", errors.ErrSvcNotFound{Service: types.NamespacedName{Namespace: namespace, Name: name}}
+	}
+	svc := svcObj.(*api_v1.Service)
+
+	// Determines whether targetPort is a number or a name.
+	// If it is a number it will need to be matched with the service port.Port.
+	// If it is a name it will need to be matched with the service port.Name.
+	portNumber := false
+	i, err := strconv.Atoi(targetPort)
```
Comment (swetharepakula) [inline pkg/controller/translator/translator.go:587]: Do we have to handle a case when target port is something like "80" and will only match a service port.Name.
Such as when `targetPort = "80"` and the matching service is `service port= { name:"80", port=81}`

**Comment by alan-kut** [comment]
Closed as I cannot reproduce the bug. It just WAI

File: pkg/controller/translator/translator.go (hunk 2 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go]: Fixed.

File: pkg/controller/translator/translator_test.go [hunk 3]
```diff
@@ -720,49 +719,262 @@ func getProbePath(p *apiv1.Probe) string {
...
+			desc:     "Port number not found",
+			svcPorts: []utils.ServicePort{createSvcPort("svc", "80", true)},
+			services: []*apiv1.Service{createService("svc", []apiv1.ServicePort{{Port: 8080, TargetPort: intstr.FromInt(80), Protocol: apiv1.ProtocolTCP}})},
+		},
+		{
+			desc:     "No ports because of wrong protocol",
+			svcPorts: []utils.ServicePort{createSvcPort("svc", "80", true)},
+			services: []*apiv1.Service{createService("svc", []apiv1.ServicePort{{Port: 80, TargetPort: intstr.FromInt(80), Protocol: apiv1.ProtocolUDP}})},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (alan-kut) [reply pkg/controller/translator/translator_test.go]: This part was removed from the solution

File: pkg/controller/translator/translator.go [hunk 4]
```diff
@@ -556,6 +562,52 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
...
+	if err == nil {
+		portNumber = true
+	}
+
+	// Finds a correct service port matched either by name or number.
+	for _, port := range svc.Spec.Ports {
+		if port.Protocol == api_v1.ProtocolTCP && (portNumber && port.Port == int32(i)) || (!portNumber && port.Name == targetPort) {
```
[the remark this answers is not in the document]
Comment (alan-kut) [reply pkg/controller/translator/translator.go]: It was removed.

**Comment by alan-kut** [comment]
Current state:

Bugs:
1) For a service with a port `{Name="portName", Port=80, TargetPort="podPortName"}` the current code will look for endpoints with ports named "podPortName" but the endpoints will have a port named "portName"
2) For a service with a port `{Name="portName", Port=80, TargetPort="1000"}` the current code will treat "1000" as a port number instead of name but it should look for endpoints with ports named "portName".

Solutions:
For 1) Added `PortName string` to the `ServicePort` structure.
For 2) Changed the type of a `TargetPort` in the `ServicePort` from `string` to `intstr.IntOrString` so it is possible to distinguish between `int 1000` and `string "1000"`.

File: pkg/controller/translator/translator.go [hunk 5]
```diff
@@ -558,10 +564,6 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
 
 // returns target port if port number is specified, finds the actual target port behind the named target port
```
Comment (freehan) [inline pkg/controller/translator/translator.go:565]: update comment to match algorithm

File: pkg/controller/translator/translator.go [hunk 6]
```diff
@@ -586,11 +588,6 @@ func listEndpointTargetPortsFromEndpointSlices(indexer cache.Indexer, namespace,
...
```
Comment (freehan) [inline pkg/controller/translator/translator.go:589]: update comment to match algorithm

File: pkg/controller/translator/translator_test.go [hunk 7]
```diff
@@ -249,9 +251,62 @@ func TestGetServicePort(t *testing.T) {
...
 			},
-			id:       utils.ServicePortID{Port: v1.ServiceBackendPort{Name: "http"}},
-			wantErr:  true,
-			wantPort: true,
+			id:         utils.ServicePortID{Port: v1.ServiceBackendPort{Name: "http"}},
+			wantErr:    true,
+			wantPort:   true,
+			wantedPort: apiv1.ServicePort{Name: "http", Port: 80},
+		},
+		{
```
Comment (freehan) [inline pkg/controller/translator/translator_test.go:259]: add one case with double named reference?

[]apiv1.ServicePort{
					{Name: "http", Port: 80, , targetPort: "pod-http"},
					{Name: "https", Port: 443, targetPort: "pod-https"},
					{Name: "otherPort", Port: 12345, , targetPort: "pod-otherPort"},

File: pkg/controller/translator/translator_test.go [hunk 8]
```diff
@@ -249,9 +251,62 @@ func TestGetServicePort(t *testing.T) {
...
+				Type: apiv1.ServiceTypeNodePort,
+				Ports: []apiv1.ServicePort{
+					{Name: "http", Port: 80, NodePort: 123, TargetPort: intstr.FromString("podport")},
+				},
+			},
+			annotations: map[string]string{
+				"service.alpha.kubernetes.io/app-protocols": "bad-string",
+			},
+			id:         utils.ServicePortID{Port: v1.ServiceBackendPort{Number: 80}},
+			wantErr:    true,
```
Comment (freehan) [inline pkg/controller/translator/translator_test.go:307]: add one with valid setup?

File: pkg/controller/translator/translator.go (hunk 5 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go:565]: Done

File: pkg/controller/translator/translator.go [hunk 9]
```diff
@@ -558,10 +564,6 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
...
 func listEndpointTargetPortsFromEndpointSlices(indexer cache.Indexer, namespace, name, targetPort string) []int {
```
[the remark this answers is not in the document]
Comment (alan-kut) [reply pkg/controller/translator/translator.go:566]: Done

File: pkg/controller/translator/translator.go (hunk 6 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go:589]: Done

File: pkg/controller/translator/translator.go [hunk 10]
```diff
@@ -586,11 +588,6 @@ func listEndpointTargetPortsFromEndpointSlices(indexer cache.Indexer, namespace,
 
 // returns target port if port number is specified, finds the actual target port behind the named target port
 func listEndpointTargetPortsFromEndpoints(indexer cache.Indexer, namespace, name, targetPort string) []int {
```
[the remark this answers is not in the document]
Comment (alan-kut) [reply pkg/controller/translator/translator.go:590]: Done

File: pkg/controller/translator/translator_test.go (hunk 7 above)
Comment (alan-kut) [reply pkg/controller/translator/translator_test.go:259]: Done

File: pkg/controller/translator/translator_test.go (hunk 8 above)
Comment (alan-kut) [reply pkg/controller/translator/translator_test.go:307]: Done

File: pkg/utils/serviceport.go [hunk 11]
```diff
@@ -44,8 +45,9 @@ type ServicePort struct {
 	NodePort int64
 	// Numerical port of the Service, retrieved from the Service
 	Port           int32
+	PortName       string
```
[the remark this answers is not in the document]
Comment (alan-kut) [reply pkg/utils/serviceport.go:48]: Done

**Comment by alan-kut** [comment]
Responded to all comments

File: pkg/utils/serviceport.go [hunk 12]
```diff
@@ -43,9 +44,11 @@ type ServicePort struct {
 
 	NodePort int64
 	// Numerical port of the Service, retrieved from the Service
-	Port           int32
+	Port int32
+	// Name of the port of teh Service, retrieved from the Service
```
Comment (swetharepakula) [inline pkg/utils/serviceport.go:48]: nit: the

Comment (alan-kut) [reply pkg/utils/serviceport.go:48]: Fixed

**Review by freehan — APPROVED** [review]
Thanks!
/lgtm
/approve

**Pull request merged**